File: sandbox/gumball/gumball.js

var gm = {
  model: "M1V192019",
  serial: "123997919189",
  coinValue: 0,
  cost: 0,
  count: 100,
  debug: function() {
    console.log(
    	this.model + ' ' + 
    	this.serial + ' ' + 
    	'cnt:' + this.count + ' ' +
    	'value:' + this.coinValue + ' ' + 
    	'cost:' + this.cost 
    );
  },
  insert: function(value) {
   	newValue = this.coinValue + value
   	console.log( newValue + ' > ' + this.cost )
   	if ( newValue > this.cost ) {
   		console.log( "There's already enough money!" )	
   	} else {
   		this.coinValue = newValue 
   	}
  },
  eject: function() {
	if ( this.coinValue < this.cost ) {
		console.log( "Not Enough Money!" ) 
		return 
	}
	if ( this.count > 0 ) {
		this.count-- 
		console.log( "Gumball Ejected!" ) 
		console.log( "Change Return: " + (this.coinValue - this.cost) + " cents." ) 
		this.coinValue = 0 
	}
	else {
		console.log( "Sorry! Out of Gumballs." ) 
	}
  }
} 

gm.debug()

gm.insertQuarter = function() { this.insert( 25 ) }
gm.turnTheCrank = function() { this.eject() }

gm.cost = 50
gm.debug()
gm.insertQuarter()
gm.debug()
gm.insertQuarter()
gm.debug()
gm.turnTheCrank()
gm.debug()


gm.insertCoin = function(value) { 
	this.coinValue += value 
	if (this.coinValue >= this.cost)
		this.eject()  
}

gm.returnCoins = function() { 
	console.log( "Change Return: " + this.coinValue  + " cents." ) ;
	this.coinValue = 0
}

gm.insertCoin(25)
gm.debug()
gm.insertCoin(100)	
gm.debug()

gm.insertCoin(25)
gm.debug()
gm.insertCoin(10)	
gm.debug()
gm.returnCoins()

File: sandbox/gumball/gumball.swift

class GumballMachine {
    
    var model : String = "M1V192019"
    var serial : String = "123997919189"
	var coinValue : Int = 0
	var cost : Int = 0  // configure at setup	
	var count : Int = 0

    init() {
    	self.count = 100
    }

	func insert( value : Int ) { 
		let newValue = coinValue + value
		print( newValue, " > ", cost, "?")
		if ( newValue > cost ) {
			print( "There's already enough money!" ) 
		}
		else {
			coinValue = newValue 
		}
	}	

	func eject() {
		if ( coinValue < cost ) {
			print( "Not Enough Money!" ) 
			return 
		}
		if ( count > 0 ) {
			count -= 1
			print( "Gumball Ejected!" ) 
			print( "Change Return: ", (coinValue - cost), " cents." ) 
			coinValue = 0 
		}
		else {
			print( "Sorry! Out of Gumballs." ) 
		}
	}

    func configCost( theCost : Int ) {
    	self.cost = theCost
    }
    
    func debug() {
        print( "model:", model, " serial:", serial, " cnt:", count, " value:", coinValue, " cost:", cost )
    }
    
}


/* ============== */
/* CRANK PROTOCOL */
/* ============== */


protocol CrankProtocol {

    func insertQuarter()
    func turnTheCrank()

}

extension GumballMachine: CrankProtocol {

    func insertQuarter() {
    	insert( value: 25 )
    }

    func turnTheCrank() {
    	eject()
    }

}

var m1 = GumballMachine()
m1.configCost( theCost: 50 )
m1.debug()

// Crank Test
m1.insertQuarter()
m1.debug()
m1.turnTheCrank()
m1.debug()
m1.insertQuarter()
m1.debug()
m1.turnTheCrank()
m1.debug()



/* ============== */
/* SLOT  PROTOCOL */
/* ============== */


protocol SlotProtocol {

    func insertCoin( value: Int)
    func returnCoins()

}

extension GumballMachine: SlotProtocol {

    func insertCoin( value: Int ) {
		coinValue += value 
		if (coinValue >= cost) {
			eject()      	
		}
    }

    func returnCoins() {
		print( "Change Return: ", coinValue, " cents." ) ;
		coinValue = 0    	
    }

}

var m2 = GumballMachine()
m2.configCost( theCost: 50 )
m2.debug()

// Slot Test 
m2.insertCoin( value: 25)
m2.debug()
m2.insertCoin( value: 100)	
m2.debug()



